[SYCL][Doc] Add Jan'24 release notes (#12365)
intel/llvm: sycl/ReleaseNotes.md

- Enabled `INTEL_SYCL_OPAQUEPOINTER_READY` [a47705f2d865] [9dded21c5984]
- Supported multiple SYCL targets with `native_cpu` in the same compiler invocation [aab5d747c46e]
- Made `-fintelfpga` flag imply `-fsycl`. [ad2f959d8b88] [6d48bc13c456]
- Ensured debug format for SYCL device on Windows is the expected format [73bba79a3ffb]
- Disabled parallel for range rounding at `-O0` [2c117d7a6104]
- Fixed missing optnone attribute in -O0 mode on non-FPGA target to improve debuggability [993fe5b31b43]

### SYCL Library
- Fixed sycl::vec ABI inconsistencies and made it trivially copyable [531aabfdee91] [f4ed132f243a]
- Supported missing types with ldg extension [aec8a35788c1]
- Stopped throwing exception when passing empty accessor to `handler::require()` [7bf1f57bc26e]
- Moved `addc()` and `subb()` functions out of experimental namespace [9e3852bc1c2e]
- Moved BFN function from experimental namespace [c9da5c7ff7e8]
- Improved error message when `queue::memcpy` receives a null pointer  by reporting source code info(file/line number) [626c4b5078c1]
- Improved accuracy of double `tanpi` host implementation [041a74c0cbda]
- Added more support for AMD subgroup functions [288aeaef03a9]

> - `annotated_ptr` API fixes [954730e7a823] [fbad42a398ef] [43336a15fcaf]

Reviewer: This is a little vague. Is it possible to briefly describe what the fixes are?

- Implemented info::event_profiling::command_submit support for platforms that use OpenCL version less than 2.1. [7cfa951303eb]
- Fixed required alignment in buffers [093dae1f7214]
- Improved error message when using `sycl_ext_oneapi_graph` features that are not yet implemented [8d8d3f4f52b0]
- Supported `get_multi_ptr` in ESIMD context [2f0f3814c541]
- Accept integral type in parallel_for with offsets [ecfc49c825c6]
- Added support for `assert()` in ESIMD kernels [545a043ebaf5]
- Added missing Intel archs to [`sycl_ext_oneapi_device_architecture`](https://github.com/intel/llvm/blob/c4db251179dcd1b30a1043d4780634dbe39b3e3a/sycl/doc/extensions/experimental/sycl_ext_oneapi_device_architecture.asciidoc) [c4db251179dc]
- Enabled range rounding for unnamed lambdas [4f30e6648d9e]
- Fixed loading of plugins when path to them contains non-ANSI symbols [5c30815d4821]

> - Improved error message the CGF object parameter accepts the wrong type of parameter [5801970585b2]

Author: Is "Improved the error message when the command group function object passed to `queue::submit` has the wrong signature" better?

> - Improved error message the CGF object parameter accepts the wrong type of parameter [5801970585b2]

Reviewer: Much better, yeah. Thanks!

- Improved `handler::fill` performance on 2D and 3D buffers [0249a55f810e]
- Added defaults for streaming and register_map fpga_kernel_properties [23a6f389c1e4]
- Moved GRF property conflict check to compile time [4fb92d35c7be]
- Added error when using invoke_simd with callables with struct arguments [03c13fd53a24]
- Added missing `rintf` for CUDA [3c327c736ced]
- Optimize kernel launch on CUDA by caching the max local mem size so that CUDA driver is called less [1c892f952cae]
- Added support for local accessors to lsc ESIMD API [9dce2d227b31]
- Made `sycl_ext_oneapi_graph` functions thread-safe [c8c64a673eeb]

> - Improveed diagnostic for nd_item passed to parallel_for with range [e600d7922489]

Reviewer 2: ```suggestion
- Improved diagnostic for `nd_item` passed to `parallel_for` with range [e600d7922489]
```

- Specify an error in `sycl_ext_oneapi_graph` if an executable graph is submitted to a queue with a different device or context [2c53815c6c1e]

## Bug Fixes
### SYCL Compiler
- Fixed backend option passing issues with multiple targets [12ca2db0e30b]
- Fixed `-fgpu-rdc` option for CUDA [f7595ac7527e]
- Fixed triple in offload mismatch warning [0fd9a4eb17a2]
- Fixed a compilation issue when `-enable-global-offset=false` for AMDGU and NVPTX targets [00cf4c29740b]

### SYCL Library
- Fixed stack corruption in `sycl::modf` for `sycl::half` type [7c6541dc18e3]
- Fixed `nextafter(half)` precision on host [1f10f3588e8c]
- Fixed identity values for (sub)group collectives min/max for AMD and CUDA [248389371932]
- Corrected exception thrown when kernel is submitted with a large (unsupported) `reqd_work_group_size` attribute [25c3666dffa9]
- Implemented hip `prefetch`/`mem_advise` (noop) for rocm < 5.0 [7e4cc1113987]
- Adjusted `range` definition so it has a default constructor in accordance with the SYCL spec[0289d2ac94f9]

> - Missing return statement in the `multi_ptr<void..>` conversion operator [836469ba88e2]

Reviewer 2: ```suggestion
- Added missing return statement in the `multi_ptr<void..>` conversion operator [836469ba88e2]
```

> - Optimize kernel fusion implementation so that some resources can be released earlier [a33b0a950ae6]

Reviewer 2: ```suggestion
- Optimized kernel fusion implementation so that some resources can be released earlier [a33b0a950ae6]
```

- Fixed an integer overflow that happened when sub-buffer size exceeds `INT_MAX` [2fb0b140ad3b]
- Fixed `std::enable_if` typo for `marray` and `multi_ptr` [0ef25ec3bdf8]
- Fixed integer type overflow in `SYCLDeviceRequirements` [cee07d3d2652]
- Fixed UB in group sort scratch memory access [ed4c01301a7c]
- Use primary context in HIP adapter [d1c92cb95b29]
- Allowed passing default constructed accessors to kernels [fb8ddbf7b0ea]
- Stopped `sycl_ext_oneapi_graph` subgraph nodes from creating incorrect dependencies on subsequent usages of the subgraph [92ddf8d562d7]
- Fixed ambiguous `log2` call in host code [12fa0af28e88]

### Documentation
- Updated `device_image_scope` property  in `sycl_ext_oneapi_device_global` [be8e031cd078]

## API/ABI breakages

> - Deprecate experimental functions for ESIMD: quot,div,mod,dp*,line,lrp [76976a22ba2e]
> - Remove ESIMD Emulator plugin [1ece6da6b113]
> - Deprecate `sycl::abs(genfloat)` [d15d4409b2cf]

Reviewer 2: ```suggestion
- Deprecated experimental functions for ESIMD: quot,div,mod,dp*,line,lrp [76976a22ba2e]
- Removed ESIMD Emulator plugin [1ece6da6b113]
- Deprecated `sycl::abs(genfloat)` [d15d4409b2cf]
```

# Nov'23 release notes
Release notes for commit range f4e0d3177338..f4ed132f243a

## New Features
### SYCL Compiler

> - Added `-fpreview-breaking-changes` option [08febcfcb64a] [99b97eef4278]

Reviewer 2: @steffenlarsen, @stdale-intel, should we elaborate here a bit more about the option purpose? In any case we can probably add a reference to a section where we describe changes we did under that flag

- Added [`sycl_ext_intel_fpga_mem`](https://github.com/intel/llvm/blob/a162179381fd1b0ebab5cab8978e7cef9dfdac8c/sycl/doc/extensions/proposed/sycl_ext_intel_fpga_mem.asciidoc) extension spec [a162179381fd]
- Added [`sycl_ext_intel_cache_controls`](https://github.com/intel/llvm/blob/deb020bf740b1d34d71189f5900b1558934a09d7/sycl/doc/extensions/proposed/sycl_ext_intel_cache_controls.asciidoc) extension spec [deb020bf740b]
- Added [design doc](https://github.com/intel/llvm/blob/1713ed19f71c4401b2ff62dcbc32a8e71b336701/sycl/doc/design/CommandGraph.md) for `sycl_ext_oneapi_graph` extension [1713ed19f71c]

## Improvements
### SYCL Compiler
- Added diagnostics if SYCL-specific command line options are used without `-fsycl` [bf128c873e74]
- Enabled `INTEL_SYCL_OPAQUEPOINTER_READY` [a47705f2d865] [9dded21c5984]
- Supported multiple SYCL targets with `native_cpu` in the same compiler invocation [aab5d747c46e]
- Made `-fintelfpga` flag imply `-fsycl`. [ad2f959d8b88] [6d48bc13c456]
- Ensured debug format for SYCL device on Windows is the expected format [73bba79a3ffb]
- Disabled parallel for range rounding at `-O0` [2c117d7a6104]

> - Fixed missing optnone attribute in -O0 mode on non-FPGA target to improve debuggability [993fe5b31b43]

Reviewer 2: ```suggestion
- Fixed missing `optnone` attribute in `-O0` mode on non-FPGA target to improve debuggability [993fe5b31b43]
```

- Fixed `std::enable_if` typo for `marray` and `multi_ptr` [0ef25ec3bdf8]
- Fixed an issue with `reqd_work_group_size` attribute values bigger than `INT_MAX` being incorrectly handled [cee07d3d2652]
- Fixed UB in group sort scratch memory access [ed4c01301a7c]
- Use primary context in HIP adapter [d1c92cb95b29]
- Fixed a crash when default-constructed accessor is passed to a kernel [fb8ddbf7b0ea]
- Stopped `sycl_ext_oneapi_graph` subgraph nodes from creating incorrect dependencies on subsequent usages of the subgraph [92ddf8d562d7]
- Fixed ambiguous `log2` call in host code [12fa0af28e88]

### Documentation
- Updated `device_image_scope` property  in `sycl_ext_oneapi_device_global` [be8e031cd078]

## API/ABI breakages
- Deprecated experimental functions for ESIMD: `quot`, `div`, `mod`, `dp*`, `line`, `lrp` [76976a22ba2e]
- Removed ESIMD Emulator plugin [1ece6da6b113]
- Deprecated `sycl::abs(genfloat)` [d15d4409b2cf]

## Upcoming API/ABI breakages
- Fixed sycl::vec ABI inconsistencies and made it trivially copyable if `-fpreview-breaking-changes` is specified  [531aabfdee91] [f4ed132f243a] 

> - Updated `sycl.hpp` to not include `<complex>` and `<cmath>` if `-fpreview-breaking-changes` is set. [6ed0ab887d18] [d22c9c58d2e7]

Reviewer 2: I think we could have a standalone paragraph under this section saying that those changes are only in effect if `-fpreview-breaking-changes` flag is passed. This would allow to simplify wording of each item in the section.

> # Jan'24 release notes

Author: Hm, yea, I was following the last release's pattern, which used the month the notes were added, but it seems the previous releases use the month of the last commit as opposed to month the notes were added. I'll switch it to Nov '23 release notes then.

- Added SYCLcompat library [bedd818eb281f92b0d2d40cb6423235743161b65] [465aa56ee5b3e6b9a82d93cc992a31ec53cc7f04] [75b31a4c24b476c50da474ae8e3d0429f14e9799] [4dfaf20884c7268342c9ff57960c2d7d54f941af] [47d84897a87b8676b0a16b6e7b38b33f47cdcf99] [8c07a840311ab7b1768a487cae3a97bef9d4f9f5] [ab578107b6f8fb5de08459777fa953bd9ab9b827] [ac5a90e0c1b57030fa8bc486d191906d8f7a2c63] [a733fa74b583d60c9e20145ee77efdd6f2105da2] [c39c2138f5a0eb6b85a2deeecf9b929530b7cb18]
- Implement sycl_ext_oneapi_auto_local_range [a6d4c2120b5c41a531559b357034117d5c323c25]
- Implement sub_group_mask version 2 [3bd09b9089c5221f72f19ab1adde19da909661a7]
- Implement ext_oneapi_queue_priority [0c33fea5fac067066cc942ee6f98612700e05d57]
- Implement opt kernel feature for joint_matrix and joint_matrix_mad [946b1f2361670c7ac24e342ee77b1f467778cf94] [c38dfcf64681b3498e9658de97ed1d2c63b20165]
- Implement sycl_ext_oneapi_prefetch [e7139b0f15d0b5e9ebd8ea36c3982eb60fc1d42f] [0229456e19d5cc1771eb4e5a0a8f2d07afa4103f] [b5d69df53620a1c9d33f7575cd3e6dd1ade54508]
- Add support for querying device architecture [1ad69e59eb13245cda5c58b04db0bb79f35a597b]

> - Add esimd as an optional device feature [a5f471ea1535ec57417d403062b43a29c4df3cf4]

Reviewer 3: From the user POV we could probably say 'Improved error message when running ESIMD code on unsupported devices'

- Added `ext_oneapi_non_uniform_groups` aspect [de92299c2c09d626dcbc633c83e259d828220d03]

### Documentation
- Update `sycl_ext_oneapi_complex` from proposed to experimental [ea6aea37c6dd777ef8df68e45a93646be2ad9558]
- Add SYCLcompat documentation [180a92ad707bd35df9e98c1474dc52a1e9b3dead]
- Add draft forward progress extension [67461148b16938a318796f639ee857b02831c38d]
- Add `sycl_ext_oneapi_kernel_compiler` extension spec [5e64f96d20c22a6c2a1600515160ddf248637641]
- Add `sycl_ext_oneapi_enqueue_functions` extension spec [ebc3ddb9a60bad6f1f0e331da4044000e3a7fb60]
- Add `sycl_ext_oneapi_graph_fusion` extension spec [8c4cad45faba815ad252de895a2a1434d8c0bd1a]
- Add `sycl_ext_intel_fpga_datapath` extension spec [a162179381fd1b0ebab5cab8978e7cef9dfdac8c]
- Add `sycl_ext_intel_cache_controls` extension spec [deb020bf740b1d34d71189f5900b1558934a09d7]

> - Add design doc for SYCL-Graph Extension [1713ed19f71c4401b2ff62dcbc32a8e71b336701]

Reviewer 4: ```suggestion
- Add design doc for `sycl_ext_oneapi_graph` extension [1713ed19f71c4401b2ff62dcbc32a8e71b336701]
```

## Improvements
### SYCL Compiler
- Diagnose SYCL options in non-sycl mode. [bf128c873e74b4c991712249a3f2145264725888]

> - Modify SYCL offloading flow to use clang-linker-wrapper [d7a1a4146eb2f35ac9d40db9db671e22291f4320] [7b907cbcaea9b24b26a8beffd9a4f7afef5aecca]

Reviewer 5: Agree with @AlexeySachkov here. This mention might end up misleading users. We should be able to announce this in the next release. But. it is too early to report anything right now. Thanks @AlexeySachkov

> - Modify SYCL offloading flow to use clang-linker-wrapper [d7a1a4146eb2f35ac9d40db9db671e22291f4320] [7b907cbcaea9b24b26a8beffd9a4f7afef5aecca]

Author: Sounds good, I removed it.

> - Enable INTEL_SYCL_OPAQUEPOINTER_READY [a47705f2d8659d38a1e494b51fe3fb361e891bda] [9dded21c59849f6ab8867dca3c40e84a085b594d]

Reviewer 6: Yes, it is not exposed to the end user. But ReleaseNotes might target developers too, so I think it helps to at least mention that we have enabled Opaque pointers for all components as well.

- Support multiple SYCL targets with native_cpu in the same compiler invocation [aab5d747c46ec216e9f85868ede744d0bb35a307]
- Make -fintelfpga flag imply -fsycl. [ad2f959d8b887768b63a53225c4f5ee2b39a0f76] [6d48bc13c4569475508d43d797778cbcd314c290]

> - Adjust debug format for SYCL device on Windows [73bba79a3ffb2ee92c3f7eeff812924bafb39bea]

Reviewer 7: Maybe "Use expected debug format for SYCL device compilations on Windows"

> - Adjust debug format for SYCL device on Windows [73bba79a3ffb2ee92c3f7eeff812924bafb39bea]

Author: I went with "Ensured debug format for SYCL device on Windows is the expected format"

- Disable parallel for range rounding at -O0 [2c117d7a61041fcb06c0e42715fcc81459e4bb8c]
- Enable SPV_INTEL_optnone extension for non-fpga targets [993fe5b31b43911a903463e414f1daf1ab81337c]

> - Relax LowerESIMDSlmReservation assert [7897517c3a5ebbe4e05977d9b8cc0dd28a192d98]

Reviewer 3: Agree, we can remove this

### SYCL Library

> - vec abi unification and trivially copyable [531aabfdee91e0055e3d28459432dfa59b17850d] [f4ed132f243ab43816ebe826669d978139964df2]

Reviewer 8: I would probably go with something along the lines of "Fixed sycl::vec ABI inconsistencies and made it trivially copyable."

Should we mention it is behind the preview option?

> - vec abi unification and trivially copyable [531aabfdee91e0055e3d28459432dfa59b17850d] [f4ed132f243ab43816ebe826669d978139964df2]

Reviewer 2: I think that everything that is not enabled by default, should be mentioned separately. Perhaps we need to include those kinds of changes into "Breaking changes" section, or even to a dedicated "Upcoming breaking changes" section where we explain what else we broke under `-fpreview-breaking-changes` flag. Thoughts?

> - vec abi unification and trivially copyable [531aabfdee91e0055e3d28459432dfa59b17850d] [f4ed132f243ab43816ebe826669d978139964df2]

Author: I think the "Upcoming API/ABI breakages" is a good idea, I added that as a section and included this as an item

- Support missing types with ldg extension [aec8a35788c1953c6ae74ab1e6eb924fa09d5e4e]
- Stop throwing exception when passing empty accessor to handler::require() [7bf1f57bc26e11868d2ef136b6e5ab18ea5d519c]
- Move addc() and subb() functions out of experimental namespace [9e3852bc1c2e52e572a7617f924d548f80300222]
- Move BFN function from experimental namespace [c9da5c7ff7e80b193491932ebd9c66ee0777d1f2]
- Report source code info(file/line number) where null pointer exception occurred in memcpy [626c4b5078c14e90a66a692ac306d98065ece035]
- Improve accuracy of double tanpi host implementation [041a74c0cbda14fc8e03e3944367a8b0107656b6]

> - Add more support for AMD subgroup functions [288aeaef03a9100bfd927aab641730da6832241d]

Author: "Added support bitwise and/or/xor and logical and/or subgroup collective functions for AMD"

- annotated_ptr API fixes [954730e7a8232087f9bca19b26a7c2f8d5bf08f3] [fbad42a398efca3aa04bdba6504885abc9b333f7] [43336a15fcafec1b3c3e94753023304ec6ef0ed3]

> - Implement fallback path for profiling info support [7cfa951303eb828836637c21518ed0f1d27ad753]

Reviewer 9: This statement can be used as description of these changes:
"Implemented info::event_profiling::command_submit support for platforms that use OpenCL version less than 2.1."

- Fix require alignment in buffers [093dae1f72146ef0f640d2e667576811245eb078]

> - Throw an exception when unsupported features used in a graph [8d8d3f4f52b0c7710bafdb6acd7d9157c6a0b7d4]

Reviewer 4: ```suggestion
- Throw an exception when `sycl_ext_oneapi_graph` features not yet implemented are used [8d8d3f4f52b0c7710bafdb6acd7d9157c6a0b7d4]
```

> - Throw an exception when unsupported features used in a graph [8d8d3f4f52b0c7710bafdb6acd7d9157c6a0b7d4]

Reviewer 4: thanks, left some minor suggestions

- Support get_multi_ptr in ESIMD context [2f0f3814c5415068ad77ae954580b4a7618cccdd]
- Accept integral type in parallel_for with offsets [ecfc49c825c665047ada8be42ccb3202e4111227]
- Support assert() in ESIMD kernels [545a043ebaf52976b4a3c7bf35e2ea637428c505]
- Add missing Intel archs to sycl_ext_oneapi_device_architecture [c4db251179dcd1b30a1043d4780634dbe39b3e3a]
- Enable range rounding for unnamed lambdas [4f30e6648d9e38fefcc1f606638ea6d6604390bb]
- Replaced LoadLibraryA() with LoadLibraryEx() [713442327d4038d5cf2158c1d3d85920a30d61cf]
- Don't use legacy ANSI-only Windows API for loading plugins [5c30815d48211565812adb68f95d9b6fb788a39e]

> - Use SFINAE to reject wrong signature of CGF object parameter [5801970585b2724dae8367470152304cd2718e3c]

Reviewer 10: Improve error reporting when submitted command group object has wrong/incompatible signature.

- fill() use native functions instead of custom kernels [0249a55f810eedc42c3571cff176c5ab01606793]
- Add defaults for streaming and register_map fpga_kernel_properties [23a6f389c1e45df077c6f15b691835b2976fda4d]
- Move GRF property conflict check to compile time [4fb92d35c7beac85913edd23197c2e6506e49240]
- Add error when using invoke_simd with callables with struct arguments [03c13fd53a2413de41b68fd03d7ad7e7ef905215]
- Add missing rintf [3c327c736ced3690b283f04302146ac952f53726]

> - Cache the max local mem size in CUDA adapter [1c892f952cae7f0cea5c446534441478948e3f7d]

Reviewer 2: For users, this should probably be rewritten as "optimized something..something". Tagging @hdelan

Author: I went with "Optimize kernel launch on CUDA by caching the max local mem size so that CUDA driver is called less"

> - Add support for local accessors to lsc API [9dce2d227b31fdbba09a044500a0ea849cdeb747]

Reviewer 2: We should mark that this one is ESIMD specific. Tagging @fineg74

> - Replace USM allocator with UMF abstractions in Level Zero backend [7b9b8ce4e4c5be5503b22e4cedf341b4120d1ad6]

Reviewer 2: I'm not sure if that should be mentioned for end user. I just don't know if that has any effect on user or exposed to end users in any way. Tagging @igchor and @kswiecicki for comments

Reviewer 11: I agree. For now there are no user-visible effects (the allocation goes through the same memory pooling mechanism, the implementation just resides in UMF now).  In the future, we plan to link dynamically with UMF, which will have some consequences, but we can mention it then.

> - Make SYCL-Graph functions thread-safe [c8c64a673eebfede0d9dbc0728a4eaaa5bde0b10]

Reviewer 4: ```suggestion
- Make `sycl_ext_oneapi_graph` functions thread-safe [c8c64a673eebfede0d9dbc0728a4eaaa5bde0b10]
```

- Improve diagnostic for nd_item passed to parallel_for with range [e600d7922489635290fd2cb63300d865fc72ee72]

> - Implement missing exceptions defined by SYCL-Graphs specification [77b794bcaad6a33ff5734d938412f61b6b7475f0]

Reviewer 4: ```suggestion
- Implement missing `sycl_ext_oneapi_graph` invalid usage exceptions  [77b794bcaad6a33ff5734d938412f61b6b7475f0]
```

- Implemented supported make_device, make_queue, make_event interop functions for HIP backend [5e9d07b14a47c2097764b148c3b1dcdb9fcbbc1c]
- Add missing nearbyint for CUDA [0ef26d3ee8086cb51d4f5deb9ad8c3d81cdd8f16]
- Implement runtime query for matrix combinations [a09491f4e389dbef29128bda69a7954e4b887af6]
- Support basic kernel launches over 64-bit global ranges [192c3016e76b36b755bd13b1d32446071812c600]
- Clarify runtime error for nested kernel submissions [7afc2d03fd0040177009967aca057cc80708747b]

> - Improve handling of spec constants with padding in the middle of structs [292309c826a3e11b8cb225667b316d8ce01e2779]

Reviewer 2: We can also probably say "fix compiler crash when spec constant is an unpacked struct with padding in the middle" - that is probably more descriptive for an end user than just "improve handling"

> - Add support for CUDA backend with graph extension [367b662a49baad1a881696448399d68f491452e4]

Reviewer 4: ```suggestion
- Add support for CUDA backend with `sycl_ext_oneapi_graph` [367b662a49baad1a881696448399d68f491452e4]
```

> - Add support for specialization constants with graph extension [fca7f5e2e3fb558e613c5daa0eb547b8e25cc524]

Reviewer 4: ```suggestion
- Add support for specialization constants in `sycl_ext_oneapi_graph` nodes [fca7f5e2e3fb558e613c5daa0eb547b8e25cc524]
```

- Improve `vec::convert` implementation to use a single instruction instead of per-element [f634cc9bf799d92029da2b3090db4563191c1f0a]

### Documentation
- Update sycl_ext_oneapi_local_memory [457251feceb81845de962383f227e14522d4cf96]
- Add get-coord API and general query example to `sycl_ext_intel_matric` [579f1dec0b5b2ecd177dfb95c3d1522f57079cc6] [38ac2124c1e8034c37056256f9864e475128ad49]
- Update design doc for new offloading model [115808f5128f33321c68b9dead4fcd25f6a2ed49]

> - Update immediate command list usage [ad973f3383e265a1e209b61559e24cec37c5d111]

Reviewer 2: Tagging @rdeodhar to see if we can have a more descriptive text here

Author: "Updated description of the default value of the environment variable `SYCL_PI_LEVEL_ZERO_USE_IMMEDIATE_COMMANDLISTS` to reflect actual behavior"

- Add overloads with sycl::queue for some functions in `sycl_ext_oneapi_bindless_images` [4d0c552985bd0e4fc16740a0fc89393ef7fe3d04]

> - Add queue requirement when replay a graph [2c53815c6c1e3f8e40c5b6dbe3e5dc2aebd4fa37]

Reviewer 4: ```suggestion
- Specify an error in `sycl_ext_oneapi_graph` if an executable graph is submitted to a queue with a different device or context [2c53815c6c1e3f8e40c5b6dbe3e5dc2aebd4fa37]
```

## Bug Fixes
### SYCL Compiler
- Fix backend option passing issues with multiple targets [12ca2db0e30b155c7d6bf14320dd1ba468ac9294]
- Fix -fgpu-rdc option for CUDA [f7595ac7527e67a59e33550851366db7d886ac3f]

> - Fix triple in offload mismatch warning [0fd9a4eb17a2db0fed151ba9d975804322966bcc]

Reviewer 2: This was a small bugfix/adjustments for #9631, which was later reverted in #10950 and I'm not sure if it has ever been re-introduced. Tagging @asudarsa, @npmiller here

> - Fix a bug when -enable-global-offset=false for AMDGU and NVPTX targets [00cf4c29740b2ec8d027e59e49e376a211599a58]

Reviewer 2: ```suggestion
- Fix a compilation issue when `-enable-global-offset=false` for AMDGU and NVPTX targets [00cf4c29740b2ec8d027e59e49e376a211599a58]
```

### SYCL Library
- Fix stack corruption in sycl::modf for sycl::half type [7c6541dc18e324e36853b473a15e8df4edf97755]

> - Fix nextafter with half on host [1f10f3588e8ccab79e42eee7fd640153d701b08c]

Reviewer 2: ```suggestion
- Fix `nextafter(half)` precision on host [1f10f3588e8ccab79e42eee7fd640153d701b08c]
```
?
Tagging @steffenlarsen

Reviewer 8: Calling it the precision is a little iffy, but it is more descriptive than the old message or anything with an appropriate length, so let's roll with it.

> - Fix identity values for (sub)group collectives min/max [2483893719324831eebce8c188aed688a8a04419]

Reviewer 2: This is AMD- and CUDA-specific, we should probably clarify that. Tagging @jchlanda

> - Improve handling of large reqd_work_group_size values [25c3666dffa935c64db513d0db25205e0400c6b7]

Reviewer 2: Corrected exception thrown when kernel is submitted with a large (unsupported) `reqd_work_group_size` attribute

> - Fix hip prefetch/mem_advise (noop) for rocm < 5.0 [7e4cc11139871da9e4662e535d4335b823b2118e]

Reviewer 2: ```suggestion
- Implemented prefetch/mem_advise (noop) for rocm < 5.0 [7e4cc11139871da9e4662e535d4335b823b2118e]
```
?
Tagging @JackAKirk

Reviewer 12: The suggested change is fine.

> - Fix ranges to allow default constructor to 0s [0289d2ac94f9692e2067a3c32fb290169d84cbee]

Reviewer 2: Adjusted `range` definition so it has default constructor in accordance with the SYCL spec?

> - Fix propagation of aspects from global variables [fb368a13044897d8ac693f72c442fe1f257c05d2]

Reviewer 2: Note: SYCL kernels cannot use global variables that aren't `const` or `constexpr` ([5.4. Language restrictions for device functions](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:language.restrictions.kernels)), so there shouldn't be that many cases where it didn't work since `const` and `constexpr` values are likely to be propagated directly into the kernel.

We can say something like "fixed a few cases where optional kernel features used by kernels wouldn't be properly detected". However, it seems to me that this was a bugfix after opaque pointers enabling and therefore this issue wasn't present in previous release. If that is the case, then we can probably omit this line completely

Author: Makes sense, I removed it

- Missing return statement in the multi_ptr<void..> conversion operator [836469ba88e292ce8dd662074117881c3fdd10d7]

> - Fix `convert` to `sycl::vec<bool, N>` [9dfaf274e2cbe42e992be0519fa7d2b0eb0face5]

Reviewer 2: This was a follow-up fix from `convert` refactoring ("Improve `vec::convert` implementation to use a single instruction instead of per-element [f634cc9bf799d92029da2b3090db4563191c1f0a]") which happened in the same release and therefore, we should probably omit this line.

> - Ensure queue_impl is not unnecessarily referenced in kernel fusion [a33b0a950ae679423aff505fdbd4a2cae77241eb]

Reviewer 2: We can probably reword this into something like "optimize kernel fusion implementation, so that some resources can be released earlier". Tagging @Naghasan

> - Fix int overflow for OffsetInBytes in accessors [2fb0b140ad3be544552a1308fe49ce491bd2881a]

Reviewer 2: Fix an integer overflow which happens when sub-buffer size exceeds `INT_MAX`. Tagging @KseniyaTikhomirova here, perhaps she has more info about how exactly the issue manifested itself (was it a crash, or an incorrect exception?)

> - Fix invalid discard_write in atomic_memory_order_seq_cst [f69dae7972062d90f804c65996d31f246d2288b3]

Reviewer 2: This is an E2E test fix, it has nothing to do with the actual product, we should remove this line

- Fix std::enable_if typo for marray and multi_ptr [0ef25ec3bdf8497dbe272e0295708d704cbe7020]

> - Fix integer type overflow in SYCLDeviceRequirements [cee07d3d26520fcdaab33dfaf7759b59210675d4]

Reviewer 2: `SYCLDeviceRequirements` is an implementation detail unavailable to users. We should probably say something like "Fix an issue with `reqd_work_group_size` attribute values bigger than `INT_MAX` being [ignored/incorrectly handled]". Tagging @againull

- Fix UB in group sort scratch memory access [ed4c01301a7c2be9157df6410be3b72ac4b95df7]

> - Use primary context in HIP adapter [d1c92cb95b29b0bf44683cc84b9978d767e0a300]

Reviewer 2: This doesn't look like a bug fix, but more like an improvement, or even as an internal refactoring. Tagging @hdelan here

> - Allow passing default constructed accessors to kernels [fb8ddbf7b0ea3f22b6a2d7fce661e44123ebb2d5]

Reviewer 2: Since that is listed as a bugfix, then maybe we can say "Fixed a crash when default-constructed accessor is passed to a kernel"? Tagging @maarquitos14

> - Fix issues with sycl_ext_oneapi_graph subgraphs [92ddf8d562d736f6c556976dce2c099e6aef591e]

Reviewer 4: ```suggestion
- Stop `sycl_ext_oneapi_graph` subgraph nodes from creating incorrect dependencies on subsequent usages of the subgraph [92ddf8d562d736f6c556976dce2c099e6aef591e]
```

> - Fix ambiguous log2 call in host code

Reviewer 2: Missing commit hash

> - Add device copyable trait for annotated_arg [9f65e61a2dbf9e9b631159a5ac2ec92137996f62]

Reviewer 2: Looks more like an improvement because it was accompanied by a spec change.

I also think that saying something like "Made `annotated_arg<T>` implicitly device copyable if `T` is device copyable". Tagging @yug-intel

Author: Moved it to the improvements section

### Documentation
- Update `device_image_scope` property  in `sycl_ext_oneapi_device_global` [be8e031cd078ea0298a773aa5fc743c18e50fe97]

## API/ABI breakages

> - Deprecate experimental functions: quot,div,mod,dp*,line,lrp [76976a22ba2e86059d655aa7d5f71250160f4864]

Reviewer 2: Those are ESIMD-specific, we should clarify that

- Remove ESIMD Emulator plugin [1ece6da6b1139ad29d1ed44a05759e9b5974d3e8]
- Deprecate `sycl::abs(genfloat)` [d15d4409b2cf7e023bf6f462fac5630699f1c0a7]

> - Don't include `<complex>/<cmath>` from `sycl.hpp`, unless `-fpreview-breaking-changes` [6ed0ab887d18b18209a7d1c31b8bb8240a820f79] [d22c9c58d2e7c4936158dde6056281ef2cf0cdf2]

Reviewer 2: Tagging @aelovikov-intel and @steffenlarsen here for suggestions. Strictly speaking, this change didn't happen, because it is hidden under a flag. Should we somehow specially annotate such changes as "future API/ABI breackages"?

In any case, "unless `-fpreview-breaking-changes`" -> "if `-fpreview-breaking-changes` is set"

Reviewer 8: I do think it would be best to mark such changes as being behind the preview flag somehow.

Reviewer 10: We had an approval just to drop the includes unconditionally. I just didn't push through with https://github.com/intel/llvm/pull/11528.
